docs/sdk-configuration.md

```diff
@@ -0,0 +1,492 @@
+    doSampledLogicWhileOtherServicesAlsoGetSampled();
+  }
+}
+```
+
+
+## Auto-configuration
+
+The above presents programmatic configuration for the SDK and proposeds that the core SDK has no
+other mechanism for configuration. There is no SPI nor processing of environment variables or
+system properties. There are many mechanisms for configuration, for example Spring Boot. 
+Integration with these systems becomes easier to reason about if we consider auto-configuration at
+a layer above the core SDK.
+
+### Java Auto-Instrumentation Agent
+
+Java Auto-Instrumentation Agent is the primary means of automatically configuring the SDK. It
+contains system properties, environment variables, and SPIs for allowing a user to have a fully
+setup tracing configuration just by applying the agent. In fact, the agent does not even allow a
+user to use the SDK directly, actively blocking it. Instead of having a situation where some
+configuration is automatic in the core SDK and some in the agent, we can move it all to the agent.
+The agent already exposes exporter SPIs - it can also expose SPIs for customization of the SDK
+components that are manually configured above.
+
+- We could consider having a very similar autoconfiguration wrapper artifact as an SDK extension too.
+But we would assume the core SDK is always manually configured.
+
+To allow users of the agent to apply tracing to their own code, the agent should attempt to
+instrument dependency injection to provide an instance of `OpenTelemetry` using the agent-configured
+SDK, for example it should add it to the Spring `ApplicationContext`. For cases where dependency
+injection is not available, though, there is no option but to provide access to the SDK through a
+global variable. We can expect such usage to still function correctly even if the agent is removed
+and a different configuration mechanism is used, such as manual configuration as above, or Spring
+Sleuth.
+
+### Spring Sleuth
+
+[Spring Sleuth](https://spring.io/projects/spring-cloud-sleuth) (or any similar observability-aware server framework such as 
+[curio-server-framework](https://github.com/curioswitch/curiostack/blob/master/common/server/framework/src/main/java/org/curioswitch/common/server/framework/monitoring/MonitoringModule.java)
+or internal frameworks developed by devops teams at companies) is also a mechanism for automatically
+configuring the SDK. In general, we would expect Sleuth users to not be using the java agent.
+
+Examples of how Sleuth could work are presented above in examples using `@Bean`. In particular, we
+expect it to have its own set of configuration properties - by making sure we don't implement
+configuration properties in the core SDK, only configuration layers like the agent or a possible
+configuration wrapper, we avoid the possibility of confusion by having duplicate variables (in
+practice, OpenTelemetry naming would likely be ignored and overwritten by Spring naming).
+
+## Partial SDKs
```

Comment:
I got very confused by this section. It seems like configuration duplication. What if I by mistake do both `OpenTelemetrySdk.builder().xxxxx` as above and `DefaultOpenTelemetry.builder().xxxx` as below?

Reply:
`OpenTelwetrySdk` is only present in the full SDK artifact (`-all`). So I guess it won't be there with a partial SDK?

Or is it potential confusion for full SDK users having both entry points? I don't think our repo should have any documentation related to `DefaultOpenTelemetry`. Maybe even hide it from JavaDoc. Full SDK users use `OpenTelemetrySdk` that's it, and it's all we support. Supporting partial SDK is up to the authors of partial SDK components IMO.

Reply:
The second part, yes. So you are talking here about some other (partial) implementation of the SDK, not this one?

Reply:
Not sure what you're referring to with "this one" - my hypothetical micrometer SDK?

So I am imagining an app with dependencies on `opentelemetry-sdk-trace` and `opentelemetry-micrometer` and use `DefaultOpenTelemetry`.

Or there might be a `opentelemetry-sdk-with-micrometer` which depends on the above and provides `OpenTelemetrySdkWithMicrometerBuilder`. Obtuse but just an example of what I was thinking thinking here.

Reply:
"this one" is "this repo". I mean, we provide full SDK and we don't plan to provide partial SDKs, right? Partial SDKs come from outside, right?

Reply:
Ah yeah that's what I'm thinking

Reply:
not true, though, since we'll be pulling metrics out of the "main" SDK and only providing that piece as an "experimental" signal until it's ready to go. We'll have a "partial" SDK for metrics, at least in some sense.